### ISSUE TITLE
QWether api （和风天气api）

### ISSUE THREAD

**Issue by Mikachu2333** [breezy-weather/breezy-weather#207]
### Link to publicly available documentation of the API

https://dev.qweather.com/en/docs/start/

### Available and not available data

**Available**
1000request/per day

Real-time weather
Hourly forecast(24h)
Daily forecast(3-7d)
grid weather also

Minutely-Precipitation

Warning
Indices
Air Quality
Sun and Moon

**Not available**

Time Machine
Tropical Cyclone
Solar Radiation

Please note that this source is for Chinese especially, people in other countries are suggested to use Accuweather or Openweather as substitute.

### Are you going to develop yourself the new provider for the app?

Yes, I'd like to assist but need some help as a noob maybe.

### Give a bit more details or context about this source

Reasons
Accuweather allow only 50 requests per day.
Current China data api do not support Minutely-Precipitation also

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written the name of the source as the title of this issue.
- [X] I have updated the app to **[latest version](https://github.com/breezy-weather/breezy-weather/releases/latest)** and this source is not part of it.

**Comment by papjul** [comment]
> Accuweather allow only 50 requests per day.

That’s the first time I hear about someone having a rate-limit with AccuWeather.


> Current China data api do not support Minutely-Precipitation also

It does. It was added in v4.1.0.


I was thinking of replacing "China" provider with ColorfulClouds/Caiyun as the current provider is mostly a frontend to ColorfulClouds, and ColorfulClouds supports worldwide while the current "China" does not.

Here is the link to their documentation: https://docs.caiyunapp.com/docs/intro

It has a free-tier with 10 000 calls per day.

I'm not Chinese, but someone told me it was good.
Do you know how ColorfulClouds compares to QWeather?
I would prefer a source that can support worldwide if possible.

**Comment by Mikachu2333** [comment]
> That’s the first time I hear about someone having a rate-limit with AccuWeather.

We users in China could hardly get access to the API server of AccuWeather. 
For example, at 08:00 I refreshed the app to see the latest forecast. However, due to the Firewall of our government, hundreds of requests are blocked and waited in the night of yesterday until the Firewall gives permission to send them. And, as a consequence, my account has been disabled due to having too many concurrent requests at the same time...

> It does. It was added in v4.1.0.

Truthly, the data of it is completely nonsense. It said that "距离你最近的降水在100公里以外(English: The closest precipitation to you is more than 100 kilometers away.)", and heavy rain falls outside the window... Besides, maybe it's time to update the ReadMe file...

> It has a free-tier with 10 000 calls per day.
I took a closer look at their billing strategy and found that there may have been a misunderstanding due to a mistake in your translation software. What they meant was "Each new user will have a total of 10,000 API calls available over the next six months. And regardless of whether or not these are used up, the number will be reset to zero after six months. You will need to pay for API usage after that." This is a Chinese text trap that led to the misunderstanding.

> I'm not Chinese, but someone told me it was good. Do you know how ColorfulClouds compares to QWeather? I would prefer a source that can support worldwide if possible
As listed above, although QWeather's forecasts may not be very accurate in other countries or regions, I still recommend it because its developers have a higher free API call limit and have a more comprehensive coverage of features.

**Comment by papjul** [comment]
> Please note that this source is for Chinese especially, people in other countries are suggested to use Accuweather or Openweather as substitute.

Worldwide users probably prefer other sources, but I can see it supports worldwide so I suggest not blocking it:
https://www.qweather.com/en/about/data-source

> > That’s the first time I hear about someone having a rate-limit with AccuWeather.
> 
> We users in China could hardly get access to the API server of AccuWeather. For example, at 08:00 I refreshed the app to see the latest forecast. However, due to the Firewall of our government, hundreds of requests are blocked and waited in the night of yesterday until the Firewall gives permission to send them. And, as a consequence, my account has been disabled due to having too many concurrent requests at the same time...

Oh, I didn’t know the firewall worked that way.
So it makes sense to have providers hosted in China so it doesn’t get blocked.

> > It does. It was added in v4.1.0.
> 
> Truthly, the data of it is completely nonsense. It said that "距离你最近的降水在100公里以外(English: The closest precipitation to you is more than 100 kilometers away.)", and heavy rain falls outside the window... Besides, maybe it's time to update the ReadMe file...

This text is provided by a different endpoint that works only for the city.

Minutely is a different endpoint that will help having a precipitation graph for the next 120 minutes from a longitude and latitude (more precise than whole city forecast).

Currently, I can see on AccuWeather radar map that it's raining in Zhangwu. So I added this city on China provider and I can see the "Precipitation forecast" in Hourly forecast card.

> > It has a free-tier with 10 000 calls per day.
> 
> I took a closer look at their billing strategy and found that there may have been a misunderstanding due to a mistake in your translation software. What they meant was "Each new user will have a total of 10,000 API calls available over the next six months. And regardless of whether or not these are used up, the number will be reset to zero after six months. You will need to pay for API usage after that." This is a Chinese text trap that led to the misunderstanding.

Oh, yes, it was indeed written 10 000 calls, but I thought it was per day, just like other providers do, but you are right, my mistake!

> > I'm not Chinese, but someone told me it was good. Do you know how ColorfulClouds compares to QWeather? I would prefer a source that can support worldwide if possible
> 
> As listed above, although QWeather's forecasts may not be very accurate in other countries or regions, I still recommend it because its developers have a higher free API call limit and have a more comprehensive coverage of features.

Sure, we can add it then. Since you say you’re a novice, I would recommand you wait for #180 first, this will make it much easier to implement ;)

**Comment by Mikachu2333** [comment]
> Worldwide users probably prefer other sources, but I can see it supports worldwide so I suggest not blocking it: https://www.qweather.com/en/about/data-source

Don't worry, QWeather is provided by a Chinese company and has been registered with the Chinese government, so there is no need to worry about it being blocked by the firewall.
In my opinion, if you can recommend the corresponding services from a weather provider to users based on their approximate location, it may be a good choice.

> This text is provided by a different endpoint that works only for the city.
> Minutely is a different endpoint that will help having a precipitation graph for the next 120 minutes from a longitude and latitude (more precise than whole city forecast).
> Currently, I can see on AccuWeather radar map that it's raining in Zhangwu. So I added this city on China provider and I can see the "Precipitation forecast" in Hourly forecast card.

Thanks for answering! I'll retry to add my city to confirm if it is my mistake.

> Sure, we can add it then. Since you say you’re a novice, I would recommand you wait for #180 first, this will make it much easier to implement ;)
Okay, I'll wait for it to finish and thank you for your tip!

**Comment by papjul** [comment]
> > Worldwide users probably prefer other sources, but I can see it supports worldwide so I suggest not blocking it: https://www.qweather.com/en/about/data-source
> 
> Don't worry, QWeather is provided by a Chinese company and has been registered with the Chinese government, so there is no need to worry about it being blocked by the firewall.

By “blocking”, I meant “not showing this provider to users outside China”. I would prefer not doing that. If a user from outside China wants to use QWeather, I should not stop them from doing so.

> In my opinion, if you can recommend the corresponding services from a weather provider to users based on their approximate location, it may be a good choice.

I added a “Help me choose” button in the weather source selection dialog as a link to https://github.com/breezy-weather/breezy-weather/blob/main/docs/PROVIDERS.md
This will be available in next version, see also #201

**Comment by papjul** [comment]
> > Truthly, the data of it is completely nonsense. It said that "距离你最近的降水在100公里以外(English: The closest precipitation to you is more than 100 kilometers away.)", and heavy rain falls outside the window... Besides, maybe it's time to update the ReadMe file...
> 
> This text is provided by a different endpoint that works only for the city.
> 
> Minutely is a different endpoint that will help having a precipitation graph for the next 120 minutes from a longitude and latitude (more precise than whole city forecast).
> 
> Currently, I can see on AccuWeather radar map that it's raining in Zhangwu. So I added this city on China provider and I can see the "Precipitation forecast" in Hourly forecast card.

After checking again, you were right, the text from hourly forecast is indeed from the minutely endpoint. I don’t know why it didn’t work for you back then though.

By the way, I forgot to ask, but what did you want me to update in the README.md?

**Comment by Mikachu2333** [comment]
> > Current China data api do not support Minutely-Precipitation also
> It does. It was added in v4.1.0.

Here, the ReadMe file showed that the feature of Minutely-Precipitation is not supported when using the _China_ data source, but you said that it has been added on version 4.1.0. This should be added in the file.

**Comment by papjul** [comment]
> > > Current China data api do not support Minutely-Precipitation also
> > 
> > 
> > It does. It was added in v4.1.0.
> 
> Here, the ReadMe file showed that the feature of Minutely-Precipitation is not supported when using the _China_ data source, but you said that it has been added on version 4.1.0. This should be added in the file.

The `README.md` doesn’t mention that.

If you are talking about `PROVIDERS.md`, it was added there 3 weeks ago:
![image](https://github.com/breezy-weather/breezy-weather/assets/1449159/fd4b5c53-4832-40ea-9833-8761b0f00370)
![image](https://github.com/breezy-weather/breezy-weather/assets/1449159/73f2f87c-6db3-41eb-b86a-97610d018191)

**Comment by papjul** [comment]
> > Sure, we can add it then. Since you say you’re a novice, I would recommand you wait for #180 first, this will make it much easier to implement ;)
> 
> Okay, I'll wait for it to finish and thank you for your tip!

#180 is now implemented, you can fork latest `main` whenever you want ;)

**Comment by GrimPixel** [comment]
I would like to say that it is not preferred to have it named “China”. There are other services like 心知天气, 彩云天气. Here 和风天气 is being talked about, but seeing the source code https://github.com/breezy-weather/breezy-weather/tree/main/app/src/main/java/org/breezyweather/sources/china it is provided by 小米, and the description is 北京天气、彩云天气、中国环境监测总站.

**Comment by papjul** [comment]
@Mikachu2333 for legal reasons, we will need to remove current China source which may leave Chinese users without a reliable source. I don't mean to rush you, but if anynof us can lend you a hand, we are here to help. You can start submitting your PR as draft and we can complete on parts of the code where you are stuck. Thank you!

**Comment by Mikachu2333** [comment]
Sorry to hear that. However, I have no spare time to handle it. It might be better to mark this issue with the label 'need community help'. (*꒦ິ⌓꒦ີ)

**Comment by papjul** [comment]
Not to my knowledge. Xiaomi service will be removed in next version, please use one of the many other sources or make a pull request to implement QWeather before next version, if you need a chinese source.

**Comment by itsHenry35** [comment]
> Not to my knowledge. Xiaomi service will be removed in next version, please use one of the many other sources or make a pull request to implement QWeather before next version, if you need a chinese source.

i want to help to pr, but i just don't know about kotlin and don't have time to start everything from beginning

**Comment by itsHenry35** [comment]
<img width="1041" alt="image" src="https://github.com/breezy-weather/breezy-weather/assets/58730749/c0db38d2-0841-4505-8aa8-12a6473c6991">
trying to learn and contribute, but three hours past seems like i have messed everything up. i don't really know the project structure and i think i need help with what exactly the service kotlin and resultconverter kotlin does

**Comment by itsHenry35** [comment]
> <img alt="image" width="1041" src="https://user-images.githubusercontent.com/58730749/284075555-c0db38d2-0841-4505-8aa8-12a6473c6991.png"> trying to learn and contribute, but three hours past seems like i have messed everything up. i don't really know the project structure and i think i need help with what exactly the service kotlin and resultconverter kotlin does

@papjul

**Comment by papjul** [comment]
Instructions are here: https://github.com/breezy-weather/breezy-weather/blob/main/CONTRIBUTE.md#weather-sources

**Comment by itsHenry35** [comment]
> Instructions are here: https://github.com/breezy-weather/breezy-weather/blob/main/CONTRIBUTE.md#weather-sources
> 
> Start small. Implement only weather features first.

got it, i'll probably finish this by next weekend. gonna take lessons next week and will go on with this next weekend

**Comment by itsHenry35** [comment]
location search works! 
![image](https://github.com/breezy-weather/breezy-weather/assets/58730749/ced4f79a-0d38-43ef-82ff-5a284e61b586)
gonna next implement the basic weather forecast tomorrow.

**Comment by itsHenry35** [comment]
<img width="200" alt="image" src="https://github.com/breezy-weather/breezy-weather/assets/58730749/ed80aed4-ad29-4325-b52e-719ca55913ab">
<img width="200" alt="image" src="https://github.com/breezy-weather/breezy-weather/assets/58730749/72f82dee-f3ec-4cb0-aa64-afcd5e35bbf6">
<img width="200" alt="image" src="https://github.com/breezy-weather/breezy-weather/assets/58730749/794ef5e5-2397-465d-bd91-eedf37fe81e3">

Current weather and app language works, but i have come across some problems that I really need help.
@papjul 
1. The API always returns a code 200 success response, which makes the handle of API key limit and incorrection of the program effectless, I found that I could manually return ApiLimitReachedException() but I don't know how to manually raise the exception of incorrect api key
2. QWeather has a lot if APIs to request as it splitted its functions across MANY MANY APIs, this makes the thread stuck when requesting multiples APIs one by one, I don't know how to let the progress bar appear or perhaps let it request simultaneously (for example) 
![image](https://github.com/breezy-weather/breezy-weather/assets/58730749/9a1f5e69-17f6-41b3-bc56-120d8cdee6a5)

**Comment by papjul** [comment]
> The API always returns a code 200 success response

What a bad practice from QWeather. Return 200 SUCCESS when it failed? Who was the genius behind this? This would be worth reporting later.

> let it request simultaneously

We always request in parallel, have a look at other services when you're stuck. We create an `Observable` for every request, then we zip them and do the necessary conversions with the Converter.

**Comment by papjul** [comment]
> perhaps you can check my code in the screenshot

I did, and you are processing one request at a time, that's why I said:

> We always request in parallel, have a look at other services when you're stuck. We create an `Observable` for every request, then we zip them and do the necessary conversions with the Converter.

I would focus on making it work, then error handling of the incorrect server implementation can be handled later (I can have a look myself)

**Comment by itsHenry35** [comment]
@papjul I wonder the app would fill some missing weather information directly from other sources?

**Comment by papjul** [comment]
It can fill missing daily data with hourly data, and the other way around.

Users can also choose to use secondary sources from "Edit" button on bottom, so they can have Pollen data, for example, if not supported by QWeather.

**Comment by itsHenry35** [comment]
@papjul I submitted the PR and please review it. Please tell me the things that I have done wrong or the features that are missing.

**Comment by papjul** [comment]
Hi @Mikachu2333, I started reviewing the PR from itsHenry35, however when I tried to register with my email, a dark pattern asks us for our phone number in a 2nd step, and there doesn't seem any way to bypass it. I try to make a privacy-friendly app and promoting a source that would require all Chinese users to hand out their phone numbers to this company bothers me :(

Do you know if there is some way to register for free without this constraint, or any other more privacy-friendly source? Thanks!

**Comment by qyl27** [comment]
In Mainland China, due to policy reasons, every user who uses internet service needs to bind their account and identity information like ID card or phone number. (And every phone number in Mainland China also needs to bind to identity information for now.)  
About the following PR, maybe it can provide an input box to accept the token from users?  
So we can allow users who was agreed the privacy policy of QWeather to use it.

**Comment by Mikachu2333** [comment]
> and there doesn't seem any way to bypass it. 

To comply with government regulations, individuals who wish to use API services online must register their accounts using their phone number or identity card number. Similarly, companies operating in China are also required to follow these guidelines to prevent any potential issues.

> I try to make a privacy-friendly app

We would also like to have a privacy-friendly API service, but it seems that it is not currently possible due to security considerations.

> Do you know if there is some way to register for free without this constraint, or any other more privacy-friendly source?

Maybe openweather would be the only choice for Chinese users who is strict to privacy… 

For users who doesn't care too much about privacy (such as myself), here is an officical weather api service which allow uses to use it free (20 times per day)
https://data.cma.cn/site/apidoc.html

And the Baidu api provides free services too.(5000 times per day, Recommanded)
https://lbs.baidu.com/faq/api?title=webapi/weather/base

**Comment by Mikachu2333** [comment]
> every user who uses internet service needs to bind their account and identity information like ID card or phone number.

To be precise, the policy aims to limit the developer, not normal users.

**Comment by itsHenry35** [comment]
@papjul perhaps you can leave the prebuilt api key empty and let users to fill that in the settings

**Comment by papjul** [comment]
> @papjul perhaps you can leave the prebuilt api key empty and let users to fill that in the settings

That's exactly my problem:

> promoting a source that would require all Chinese users to hand out their phone numbers to this company bothers me :(

Thank you for your detailed messages, I will review the suggestion options. In the meantime, I will keep the other chinese source.

**Comment by Mikachu2333** [comment]
> promoting a source that would require all Chinese users to hand out their phone numbers to this company bothers me :(

OpenWeather provide China weahter data

**Comment by papjul** [comment]
All sources provide data for China, including Météo-France. I thought the issue was that they weren't approved by China govt and limited by the firewall, is OpenWeather allowed? Of all sources available, I would not recommand that one as they require credit card info now (they didn't, back when it was implemented)

**Comment by itsHenry35** [comment]
> All sources provide data for China, including Météo-France. I thought the issue was that they weren't approved by China govt and limited by the firewall, is OpenWeather allowed? Of all sources available, I would not recommand that one as they require credit card info now (they didn't, back when it was implemented)

it was just not accurate and occasionally unable to connect

**Comment by papjul** [comment]
The `qweather` branch was moved to the `archive/source-qweather` tag: https://github.com/breezy-weather/breezy-weather/tree/archive/source-qweather
